com.oracle.truffle.r.runtime/src/com/oracle/truffle/r/runtime/data/RAltIntVectorData.java

```diff
@@ -0,0 +1,286 @@
+        /* TODO: && descriptor.isUnserializeMethodRegistered(); */
+    }
+
+    public AltIntegerClassDescriptor getDescriptor() {
+        return descriptor;
+    }
+
+    public RAltRepData getAltrepData() {
+        return altrepData;
+    }
+
+    public Object getData1() {
+        return altrepData.getData1();
+    }
+
+    public Object getData2() {
+        return altrepData.getData2();
+    }
+
+    public void setData1(Object data1) {
+        altrepData.setData1(data1);
+    }
+
+    public void setData2(Object data2) {
+        altrepData.setData2(data2);
+    }
+
+    @Override
+    public void setOwner(RAbstractVector newOwner) {
+        owner = (RIntVector) newOwner;
+    }
+
+    private RIntVector getOwner() {
+        assert owner != null;
+        return owner;
+    }
+
+    public void setDataptrCalled() {
+        dataptrCalled = true;
+    }
+
+    public boolean wasDataptrCalled() {
+        return dataptrCalled;
+    }
+
+    @ExportMessage
+    public NACheck getNACheck(@Cached() NACheck na) {
+        na.enable(false);
+        return na;
+    }
+
+    @SuppressWarnings("static-method")
+    @ExportMessage
+    public final RType getType() {
+        return RType.Integer;
+    }
+
+    @ExportMessage(limit = "1")
+    public int getLength(@Cached("create()") AltrepLengthNode lengthNode) {
+        return (int) lengthNode.execute(getOwner());
+    }
+
+    @ExportMessage(limit = "1")
+    public RIntArrayVectorData materialize(@Cached("create()") AltrepRFFI.AltIntDataptrNode dataptrNode,
+                                           @Cached AltrepLengthNode lengthNode) {
+        int length = (int) lengthNode.execute(getOwner());
+        long dataptrAddr = dataptrNode.execute(getOwner(), true);
+        int[] newData = new int[length];
+        NativeMemory.copyMemory(dataptrAddr, newData, ElementType.INT, length);
+        // TODO: complete=true?
+        return new RIntArrayVectorData(newData, true);
+    }
+
+    @ExportMessage
+    public boolean isWriteable() {
+        return dataptrCalled;
+    }
+
+    @ExportMessage
+    public boolean isComplete() {
+        return true;
+    }
+
+    /**
+     * May return either RAltIntVectorData or RIntArrayVectorData, depending whether this vector data has
+     * Duplicate method registered.
+     */
+    @ExportMessage
+    public Object copy(boolean deep,
+                       @Cached AltrepDuplicateNode duplicateNode) {
+        return duplicateNode.execute(getOwner(), deep);
+    }
+
+    @ExportMessage
+    public RIntArrayVectorData copyResized(int newSize, boolean deep, boolean fillNA) {
+        throw RInternalError.unimplemented("RAltIntVectorData.copyResized");
+    }
+
+    /**
+     * TODO: This method gets called from Rf_duplicate which means that it eventually should make downcall into
+     *   Duplicate. The behavior may change, so the implementation is not optimized.
+     */
```

Comment:
Hm, we should probably:
* have a `DuplicateNode` in `c.o.t.r.runtime.data.nodes`, which should implement GNU-R compatible duplication logic, `Rf_duplicate` would just delegate to that
* new message `VectorDataLibrary#duplicate`, which would return `null` if the default implementation should be used (`DuplicateNode` would piggy-back on this).
* we should call `DuplicateNode` in the same situations as GNU-R would. E.g., if builtin `xyz` calls `duplicate` in GNU-R in some circumstance, we should too.